tests/unit_tests/blas/batch/gemm_batch_usm.cpp

```diff
@@ -139,39 +134,53 @@ int test(const device &dev, int64_t group_count) {
             else
                 transb[i] = (onemkl::transpose)tmp;
         }
-        total_size_a[i] = lda[i] * ((transa[i] == onemkl::transpose::nontrans) ? k[i] : m[i]);
-        total_size_b[i] = ldb[i] * ((transb[i] == onemkl::transpose::nontrans) ? n[i] : k[i]);
-        total_size_c[i] = ldc[i] * n[i];
+        size_a[i] = lda[i] * ((transa[i] == onemkl::transpose::nontrans) ? k[i] : m[i]);
+        size_b[i] = ldb[i] * ((transb[i] == onemkl::transpose::nontrans) ? n[i] : k[i]);
+        size_c[i] = ldc[i] * n[i];
         total_batch_count += group_size[i];
+        total_size_a += size_a[i] * group_size[i];
+        total_size_b += size_b[i] * group_size[i];
+        total_size_c += size_c[i] * group_size[i];
     }
 
-    fp **a_array     = (fp **)onemkl::malloc_shared(64, sizeof(fp *) * total_batch_count, dev, cxt);
-    fp **b_array     = (fp **)onemkl::malloc_shared(64, sizeof(fp *) * total_batch_count, dev, cxt);
-    fp **c_array     = (fp **)onemkl::malloc_shared(64, sizeof(fp *) * total_batch_count, dev, cxt);
-    fp **c_ref_array = (fp **)onemkl::malloc_shared(64, sizeof(fp *) * total_batch_count, dev, cxt);
-
-    if ((a_array == NULL) || (b_array == NULL) || (c_array == NULL) || (c_ref_array == NULL)) {
-        std::cout << "Error cannot allocate arrays of pointers\n";
-        onemkl::free_shared(a_array, cxt);
-        onemkl::free_shared(b_array, cxt);
-        onemkl::free_shared(c_array, cxt);
-        onemkl::free_shared(c_ref_array, cxt);
-        return false;
-    }
+    auto uafpp = usm_allocator<fp *, usm::alloc::shared, 64>(cxt, dev);
+    vector<fp *, decltype(uafpp)> a_array(uafpp), b_array(uafpp), c_array(uafpp),
+        c_ref_array(uafpp);
+
+    a_array.resize(total_batch_count);
+    b_array.resize(total_batch_count);
+    c_ref_array.resize(total_batch_count);
+    c_array.resize(total_batch_count);
+
+    vector<fp, decltype(uafp)> a_array_data(uafp), b_array_data(uafp), c_array_data(uafp),
+        c_ref_array_data(uafp);
+    a_array_data.resize(total_size_a);
+    b_array_data.resize(total_size_b);
+    c_array_data.resize(total_size_c);
+    c_ref_array_data.resize(total_size_c);
+
     idx = 0;
     for (i = 0; i < group_count; i++) {
         for (j = 0; j < group_size[i]; j++) {
-            a_array[idx] = (fp *)onemkl::malloc_shared(64, sizeof(fp) * total_size_a[i], dev, cxt);
-            b_array[idx] = (fp *)onemkl::malloc_shared(64, sizeof(fp) * total_size_b[i], dev, cxt);
-            c_array[idx] = (fp *)onemkl::malloc_shared(64, sizeof(fp) * total_size_c[i], dev, cxt);
-            c_ref_array[idx] =
```

Comment:
Is there a reason why changing from independent pointers as allowed by the API to contiguous matrices?

Reply:
cuBLAS group API is expecting contiguous matrices. This change is done to support all backends.

Reply:
Do you have a pointer to the cuBLAS documentation that mentions this requirement? I could not find it.

Reply:
I take it back, you are right. I got confused. cuBLAS is also accepting array of pointers. 
Regarding contiguous matrices: Using malloc_shared for pointer arrays caused "CUDA_ERROR_ILLEGAL_ADDRESS" error in CUDA PI. As a workaround, I switched allocating array pointers and matrices using usm_allocator<T, usm::alloc::shared>. With that, I couldn't find a way to allocate matrices in a loop and still keep them alive beyond scope. Instead, I allocated a large array and copied the temporary matrices into this large array. This large array ended up being contiguous. In backends though, each array is accessed through pointers, so, they don't really care (or use) if the large array is contiguous or not.

Reply:
Updated memory allocation for matrices using vector of vectors.

Reply:
@lhuot Does this look good?

Reply:
After some study, I realized that, vector of vectors will also provide contiguous data array. So, I did more debug on why malloc_shared does not work in this case. The root-cause is using non-aligned memory. After fixing it, malloc_shared works fine and the non-contiguous data arrays are provided to the library calls.

Reply:
Thanks!